Vite 3.2.0 type definitions depend on picomatch

### Describe the bug

Updating Vite to 3.2.0 causes TypeScript to complain about missing `picomatch` dependency, which is dev dependency of `vite`.

```
node_modules/vite/dist/node/index.d.ts(35,43): error TS2307: Cannot find module 'picomatch' or its corresponding type declarations.
```

This can be seen on https://github.com/xfix/live/actions/runs/3332827904/jobs/5514140592.

`index.d.ts` contains the following import.

```javascript
import type { Matcher as Matcher_2 } from 'picomatch';
```

`Matcher_2` is not used anywhere.

### Reproduction

https://github.com/xfix/live/actions/runs/3332827904/jobs/5514140592

### Steps to reproduce

_No response_

### System Info

```shell
GitHub Actions, vite 3.2.0
```


### Used Package Manager

npm

### Logs

_No response_

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitejs/vite/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitejs.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitejs/vite/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a Vite issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to [vuejs/core](https://github.com/vuejs/core) instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite/discussions) or join our [Discord Chat Server](https://chat.vitejs.dev/).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/packages/vite/scripts/postPatchTypes.ts
+++ b/packages/vite/scripts/postPatchTypes.ts
@@ -1,7 +1,8 @@
 import { dirname, resolve } from 'node:path'
 import { fileURLToPath } from 'node:url'
+import fs from 'node:fs'
 import colors from 'picocolors'
-import { rewriteImports } from './util'
+import { rewriteImports, walkDir } from './util'
 
 const dir = dirname(fileURLToPath(import.meta.url))
 const nodeDts = resolve(dir, '../dist/node/index.d.ts')
@@ -14,3 +15,18 @@ rewriteImports(nodeDts, (importPath) => {
 })
 
 console.log(colors.green(colors.bold(`patched types/* imports`)))
+
+// remove picomatch type import because only the internal property uses it
+const picomatchImport = "import type { Matcher as Matcher_2 } from 'picomatch';"
+
+walkDir(nodeDts, (file) => {
+  const content = fs.readFileSync(file, 'utf-8')
+  if (!content.includes(picomatchImport)) {
+    throw new Error(`Should find picomatch type import in ${file}`)
+  }
+
+  const replacedContent = content.replace(picomatchImport, '')
+  fs.writeFileSync(file, replacedContent, 'utf-8')
+})
+
+console.log(colors.green(colors.bold(`removed picomatch type import`)))

--- a/packages/vite/scripts/util.ts
+++ b/packages/vite/scripts/util.ts
@@ -21,7 +21,7 @@ export function slash(p: string): string {
   return p.replace(/\\/g, '/')
 }
 
-function walkDir(dir: string, handleFile: (file: string) => void): void {
+export function walkDir(dir: string, handleFile: (file: string) => void): void {
   if (statSync(dir).isDirectory()) {
     const files = readdirSync(dir)
     for (const file of files) {

--- a/packages/vite/tsconfig.check.json
+++ b/packages/vite/tsconfig.check.json
@@ -2,6 +2,8 @@
   "compilerOptions": {
     "noEmit": true,
     "moduleResolution": "classic",
+    "noResolve": true,
+    "typeRoots": [],
     // Only add entries to `paths` when you are adding/updating dependencies (not devDependencies)
     // See CONTRIBUTING.md "Ensure type support" for more details
     "paths": {
@@ -14,9 +16,22 @@
       // indirect: postcss depends on it
       "source-map-js": ["./node_modules/source-map-js/source-map.d.ts"]
     },
-    "typeRoots": [],
     "strict": true,
     "exactOptionalPropertyTypes": true
   },
-  "include": ["dist/**/*.d.ts"]
+  "include": [
+    "../../node_modules/@types/node/**/*",
+    // dependencies
+    "./node_modules/rollup/**/*",
+    "./node_modules/esbuild/**/*",
+    "./node_modules/postcss/**/*",
+    "./node_modules/source-map-js/**/*",
+    // dist
+    "dist/**/*",
+    "types/customEvent.d.ts",
+    "types/hmrPayload.d.ts",
+    "types/importGlob.d.ts",
+    "types/importMeta.d.ts",
+    "types/hot.d.ts"
+  ]
 }

